python/ray/util/sgd/tests/test_torch_runner.py

```diff
@@ -203,54 +203,96 @@ def _testWithInitialized(self, init_mock):
-            args, _ = mock_runner._set_cuda_device.call_args
-            self.assertTrue(("1" in args) or "0" in args)
-            self.assertEquals(len(os.environ["CUDA_VISIBLE_DEVICES"]), 1)
+            init_mock.return_value = True
+            self._testReserveCUDAResource(init_mock, 0)
+
+    def test2VisibleWithInitializedAndReserveCPUResource(self):
+        os.environ["CUDA_VISIBLE_DEVICES"] = "2,3"
+        with patch("torch.cuda.is_initialized") as init_mock:
+            init_mock.return_value = True
+            self._testReserveCUDAResource(init_mock, 2)
 
     def test1VisibleNotInitialized(self):
         os.environ["CUDA_VISIBLE_DEVICES"] = "0"
         with patch("torch.cuda.is_initialized") as init_mock:
             init_mock.return_value = False
-            mock_runner = MagicMock()
-            mock_runner._set_cuda_device = MagicMock()
-            LocalDistributedRunner._try_reserve_resources_and_set_cuda(
-                mock_runner, 0, 1)
-            mock_runner._set_cuda_device.assert_called_with("0")
-            self.assertEquals(len(os.environ["CUDA_VISIBLE_DEVICES"]), 1)
+            self._testReserveCUDAResource(init_mock, 0)
+
+    def test1VisibleNotInitializedAndReserveCPUResource(self):
+        os.environ["CUDA_VISIBLE_DEVICES"] = "0"
+        with patch("torch.cuda.is_initialized") as init_mock:
+            init_mock.return_value = False
+            self._testReserveCUDAResource(init_mock, 2)
+
+    def test2VisibleNotInitialized(self):
+        os.environ["CUDA_VISIBLE_DEVICES"] = "2,3"
+        with patch("torch.cuda.is_initialized") as init_mock:
+            init_mock.return_value = False
+            self._testReserveCUDAResource(init_mock, 0)
+
+    def test2VisibleNotInitializedAndReserveCPUResource(self):
+        os.environ["CUDA_VISIBLE_DEVICES"] = "2,3"
+        with patch("torch.cuda.is_initialized") as init_mock:
+            init_mock.return_value = False
+            self._testReserveCUDAResource(init_mock, 2)
 
     @patch("torch.cuda.set_device")
     def testSetDevice(self, set_mock):
         mock_runner = MagicMock()
         mock_runner._is_set = False
         LocalDistributedRunner._set_cuda_device(mock_runner, "123")
-        self.assertEquals(mock_runner.local_device, "123")
+        self.assertEquals(mock_runner.local_cuda_device, "123")
         self.assertTrue(set_mock.called)
         set_mock.assert_called_with(123)
+
+    def testReserveCPUResources(self):
+        mock_runner = MagicMock()
+        mock_runner._set_cpu_devices = MagicMock()
+        # reserve CPU only
+        LocalDistributedRunner._try_reserve_and_set_resources(
+            mock_runner, 4, 0)
+
+        available_devices = set(range(0, 10))
+
+        self.assertTrue(mock_runner._set_cpu_devices.called)
+        set_devices = mock_runner._set_cpu_devices.call_args[0][0]
+        self.assertEquals(len(set_devices), 4)
+        self.assertTrue(set(set_devices).issubset(available_devices))
+
```

Comment:
OK instead of putting code into the runner, how about just checking the ray global state? `ray.available_resources()["CPU"]` should indicate that the remaining CPUs is less than the total cluster_resources.

Reply:
updated